More text on assembly and rust frontend. (#108)
chriseth/powdr: README.md

# powdr

> an extended polynomial identity language (PIL) and zk-focused assembly in rust

Reviewer: ```suggestion
an extended polynomial identity language (PIL) and zk-focused assembly in Rust.
```

WARNING: This is a proof-of concept implementation. It is missing many internal checks. DO NOT USE FOR PRODUCTION!

## Basic Concept

> powdr is a toolkit that helps build zkEVMs and similar proof systems.

Reviewer: ```suggestion
powdr is a toolkit that helps build zkVMs and similar proof systems.
```

It has two main components:

- A polynomial identity language that allows you to define polynomial constraints, lookups, etc.
- An extensible assembly language to perform dynamic executions.
  

> Both the frontends and the backends are flexible.

Reviewer: ```suggestion
Both frontend and backend are highly flexible.
```

> As an example, powdr contains a frontend that enables you to write code in (no-std) rust,

Reviewer: ```suggestion
As an example, powdr contains a frontend that enables you to write code in (no-std) Rust,
```

which gets compiled to RISCV, then to powdr-assembly and finally to PIL.

All stages of this execution are fully automatic, which means you do not need to write
additional code for witness generation. All witnesses are automatically inferred from the constraints.

> Since the witnesses are inferred, the system can also ensure that there are no additional (unwanted) witnesses.

Reviewer: `and the system is not underconstrained.`

Author: That is the word I was looking for!

All artifacts during this compilation process are human-readable. This means you can take a look at
the RISCV assembly file, the powdr assembly file and the PIL file.

The assembly language is designed to be extensible. This means that it does not have a single
native instruction. Instead, all instructions are user-defined and because of that,
it is easy to adapt powdr assembly to a different kind of VM.

### How to run the Rust-RISCV example

```sh
# Install the riscv target for the rust compiler
rustup target add riscv32imc-unknown-none-elf
# Run the compiler. It will generate files in /tmp/.
# -i specifies the prover witness input (see below)
cargo run --release rust tests/riscv_data/sum.rs -o /tmp -f -i 10,2,4,6 
```

> The example rust file verifies that a supplied list of integers sums up to a specified value
> (yes this is the full and only input file you need for the whole process):

Reviewer: ```suggestion
The following example Rust file verifies that a supplied list of integers sums up to a specified value.
Note that this is the full and only input file you need for the whole process!
```
